notebook-src/02-book-balancing-intro.rst

```diff
@@ -0,0 +1,298 @@
+
+All systems will have different sets of parameters. This system may have more
+parameters too, for example what if you were on the moon? Maybe the
+acceleration due to gravity could be changed or what if the book to cup
+connection was very slippery, maybe the coefficient of friction would be a
+specific parameter. Parameters, as we've defined here, are constant with
+respect to time.
+
+Time Varying Parameters
+=======================
+
+There are other types of parameters too. A very important type of parameter are
+those that vary with time.
+
+Exercise
+--------
+
+What are the time varying parameters of this system?
+
+
+There are are an infinite number of time varying parameters. Most systems are
+described by a uniquely simple set of time varying parameters, often called
+generalized coordinates. These coordinates define the configuration of the
+system. In our case the vertical and horizontal location of the book's mass
+center could uniquely describe the configuration of the system (if the book
+can't slip on the cup). But a better choice would be to use the single time
+varying angle of the books surface relative to horizontal to define the
+configuration. The angle of the book is a generalized coordinate because no
+fewer number of parameters can possible be used to describe the configuration.
+This number of generalized coordinates corresponds to the number of degrees of
+freedom of a system. The degrees of freedom are XXX. The non-slipping book on a
+cup has 1 degree of freedom which is described by the single generalized
+coordinate, the book's angle. The system's generalized coordinates can be
+accessed as such::
+
+   >>> sys.coordinates
+   {'book_angle': 0.0}
+   >>> sys.coordinates['book_angle']
+   0.0
+
+Another type of time varying parameter that can be extracted from systems are
+*measurement parameters*. For example, maybe you are interested in the vertical
+and horizontal location of the book's center of mass instead of the generalized
+coordinate. These two measurement parameters are a function of the book angle
+and the system's geometry. We will use Python functions to define this
+relationship. Given the value of a generalized coordinate and the values of the
+system's parameters, we can define a function that computes the measurement
+parameter. For example::
+
+   >>> import numpy as np
+   >>> def compute_vertical_mass_location():
+           # in the sys class this function will be wrapped and all of the
+           # parameters, coordinates, and measurements will be injected into the
+           # namespace just above the function so the students can just write
+           # these functions with the correct variables and not worry about
+           # unpacking arrays
+           return ((radius + height / 2) * np.cos(book_angle) + radius *
+                    book_angle * np.sin(book_angle) + radius)
+   >>> sys.add_measurement('mass_center_height', compute_vertical_mass_location)
```

Comment:
I think this is doable by having a wrapper method in the class that looks something like:

```python
def compute_measurements(self):
    l = locals()
    for k, v in self.parameters.items():
        l[k] = v
    return self.meas_funcs['mass_center_height']()
```

Reply:
This seems to work:

```python
import collections
from inspect import getargspec


class MeasurmentsDict(collections.MutableMapping, dict):
    def __getitem__(self, key):
        # call function to compute value
        func = self._funcs[key]
        args = [self._par[k] for k in getargspec(func).args]
        return func(*args)

    def __setitem__(self, key, value):
        raise ValueError('Cannot set')


class System(object):
    def __init__(self):
        self.par = {'a': 1, 'b': 2, 'c': 3}
        self.meas = MeasurmentsDict({})
        self.meas._funcs = {}
        self.meas._par = self.par

    def add_meas(self, name, func):
        self.meas._funcs[name] = func


sys = System()


def compute(a, b, c):
    return a + b + c

sys.add_meas('something', compute)

print(sys.meas['something'])

sys.par['a'] = 5

print(sys.meas['something'])
```